lib/medtronic600/NGPHistoryParser.js

```diff
@@ -618,28 +832,217 @@ class NGPHistoryParser {
+          agent: 'manual',
+        });
+
+      this.addBtUtcFields(timeChange, event.timestamp);
+
+      timeChangeEvents.push(timeChange);
+    }
+
+    this.cfg.tzoUtil = new TZOUtil(this.cfg.timezone, mostRecent, timeChangeEvents);
+
+    return {
+      postRecords: this.cfg.tzoUtil.records,
+      tzoUtil: this.cfg.tzoUtil,
+    };
+  }
+
+  buildSettingsRecords(events) {
+    // Start with current settings, and work our way backwards in time to build settings change
+    // snapshots.
+    const currentSettings = this.cfg.builder.makePumpSettings();
+    currentSettings.with_units(this.currentPumpSettings.units)
+      .with_carbRatio(this.currentPumpSettings.carbRatio)
+      .with_insulinSensitivity(this.currentPumpSettings.insulinSensitivity)
+      .with_bgTarget(this.currentPumpSettings.bgTarget)
+      .with_basalSchedules(this.currentPumpSettings.basalSchedules)
+      .with_activeSchedule(this.currentPumpSettings.activeSchedule);
+    const currentNgpTimestamp = this.currentPumpSettings.currentNgpTimestamp;
+    this.addBtUtcFields(currentSettings, new NGPUtil.NGPTimestamp(currentNgpTimestamp.rtc,
+      currentNgpTimestamp.offset));
+
+    for (const event of this.eventsOfTypeReverse(
+        NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_BG_TARGETS,
+        NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_INSULIN_SENSITIVITY,
+        NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_INSULIN_TO_CARB_RATIOS,
+        NGPHistoryEvent.EVENT_TYPE.NEW_BASAL_PATTERN,
+        NGPHistoryEvent.EVENT_TYPE.BASAL_PATTERN_SELECTED)) {
+      const settingsChange = _.cloneDeep(currentSettings);
+
+      switch (event.eventType) {
+        case NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_BG_TARGETS:
+          {
+            const targets = event.bgTargets();
+            // Because we read the initial pump settings in mg/dL, always convert settings changes.
+            if (event.units === NGPUtil.NGPConstants.BG_UNITS.MMOL_L) {
+              for (const target of targets) {
+                target.high = Math.ceil(target.high * 18);
+                target.low = Math.ceil(target.low * 18);
+              }
+            }
+            settingsChange.with_bgTarget(targets);
+          }
+          break;
+        case NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_INSULIN_SENSITIVITY:
+          {
+            const isfs = event.isfs();
+            // Because we read the initial pump settings in mg/dL, always convert settings changes.
+            if (event.units === NGPUtil.NGPConstants.BG_UNITS.MMOL_L) {
+              for (const isf of isfs) {
+                isf.amount = Math.ceil(isf.amount * 18);
+              }
+            }
+            settingsChange.with_insulinSensitivity(isfs);
+          }
+          break;
+        case NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_INSULIN_TO_CARB_RATIOS:
+          {
+            const carbRatios = event.carbRatios();
+            // TODO - Exchange to grams conversion
```

Comment:
In this particular instance, the data model doesn't support exchanges (see http://developer.tidepool.io/data-model/device-data/types/pumpSettings.html#units), but I've fixed up the mg/dL and mmol/L settings changes.